README.md
# fool-deploy

[![Test](https://github.com/thomas-void0/fool-deploy/actions/workflows/test.yml/badge.svg)](https://github.com/thomas-void0/fool-deploy/actions/workflows/test.yml)
[![Publish](https://github.com/thomas-void0/fool-deploy/actions/workflows/publish.yml/badge.svg)](https://github.com/thomas-void0/fool-deploy/actions/workflows/publish.yml) <img src="https://img.shields.io/npm/v/fool-deploy.svg" alt="npm package">

[简体中文](./README.zh.md)

You can use `fool-deploy` to quickly and easily deploy your blogs, demos, frontend pages, and more on the server.

# demo

Deploy the application created by `vite` and `create-react-app`:

![image](https://user-images.githubusercontent.com/48620706/227898197-5314717d-5184-4d64-b7c9-d6f0bfe9b8d2.png)
![image](https://user-images.githubusercontent.com/48620706/227898276-aeea45d6-9d6b-40f4-90cc-aa48b73ba420.png)

# Precondition

1. a linux server.
2. in your server [install docker engine](https://docs.docker.com/engine/install/centos/).（notes:Choose the type of installation for your server os type）
3. Through `git` or [XFTP](https://www.xshell.com/en/xftp/) or `Linux rz` command will your project source code pull or copied to the server
4. Install `node` and `npm` on your server.

# How use

1. install deps

```shell
$ npm i fool-deploy

or

$ yarn add fool-deploy

or

$ pnpm i fool-deploy
```

2. modify your package.json, add the following command:

```shell
"scripts": {
  "deploy":"fool"
}
```

3. in your web project root dir run:

```shell
$ npm run deploy

or

$ yarn deploy

or

$ pnpm deploy
```

# Configuration

if you want custom config, you need create `.foolrc` in project root dir.

```json
{
  "cache": true,
  "port": 2333,
  "packageCommand": "npm",
  "buildCommand": "npm run build",
  "nodeVersion": "18.14-alpine",
  "nginxVersion": "1.22.1",
  "imageName": "fool-deploy:prod",
  "output": "dist"
}
```

## cache

`cache`:Indicates whether to use caching; defaults to `true`.If you set`true`，The `.fool-cache` directory will be generated at the root of the project after the first build.Each subsequent build will run the configuration file in `.foolrc` with the same `.foolrc` configuration.
Using caching can shorten the time it takes to deploy your project later.

## port

`port`:Indicates the port on which the project will run after deployment; defaults to `2333`.If the port number is occupied, it will automatically `+1` until an available port number is found.

## packageCommand

`packageCommand`:Indicates the package manager used to install dependencies and build the project. By default, this will read the package manager you used to execute the deployment command.
for example you running `yarn deploy`，the option is specified as `yarn`.The configurable options are `npm` | `yarn` | `pnpm`.This option is very important，
Suppose the `npm` manager used by the project is packaged, but specified to run as `yarn` at deployment time. This could cause dependencies to fail to install or build during deployment.

## buildCommand

`buildCommand`:Indicates the project build command，defaults to specified `packageCommand` field `[packageCommand] build`.

## nodeVersion

`nodeVersion`:Indicates the version of the `node` environment required to install dependencies and packages for the deployed project.Default is `18.14-alpine`.
When this version does not meet the project packaging needs, choose a runnable version. More version information can be found at [dockerhub node](https://hub.docker.com/_/node).

## nginxVersion

`nginxVersion`:Indicates the version of `nginx` environment to run for the deployed project.Default is `1.22.1`.When this version does not meet the project packaging needs, choose a runnable version. More version information can be found at [dockerhub nginx](https://hub.docker.com/_/nginx).

## imageName

`imageName`:Indicates the name of the image that `docker` is building, and by default it reads the `name` field in `package.json`.the format is `name:prod`.

## output

`output`:Indicates the bundled output directory of the project, equivalent to the output directory specified by `webpack`. The default is `dist`.If the output directory in your project is `build`，You must specify it in `.foolrc`. Otherwise, it will cause an error that the resouce cannot be found.

# License

[MIT](./LICENSE)


CHANGELOG.md
# [0.1.0](https://github.com/thomas-void0/fool-deploy/compare/v0.0.12...v0.1.0) (2023-03-28)


### Features

* add packageCommand field ([dae2ad6](https://github.com/thomas-void0/fool-deploy/commit/dae2ad604d3d02bd4df43684d755e91d5f7d61da))



## [0.0.13](https://github.com/thomas-void0/fool-deploy/compare/v0.0.12...v0.0.13) (2023-03-27)



## [0.0.11-beta](https://github.com/thomas-void0/fool-deploy/compare/v0.0.6-beta...v0.0.11-beta) (2023-03-27)


### Bug Fixes

* fix ip address bug ([33885c5](https://github.com/thomas-void0/fool-deploy/commit/33885c5f9588eb000e051cf40aca775416d25eef))
* fix packagecommand get error ([b9a2ee6](https://github.com/thomas-void0/fool-deploy/commit/b9a2ee6cae3740808b109a270ecabab218a6d1f6))
* fix shell exec error ([4b6087f](https://github.com/thomas-void0/fool-deploy/commit/4b6087f52d8f409d3f39405d51db68d2e13da246))


### Features

* add port tip ([bbf73dc](https://github.com/thomas-void0/fool-deploy/commit/bbf73dc13db6797dbea459a0fd001a96c899322f))
* add probe port ([cfc7080](https://github.com/thomas-void0/fool-deploy/commit/cfc7080682e663a00f046718aafeac87eae9fbaf))
* add visit address console ([c2d92cf](https://github.com/thomas-void0/fool-deploy/commit/c2d92cf4d35bb53558d1a3d2c606e4edabbaa13c))



## [0.0.9-beta](https://github.com/thomas-void0/fool-deploy/compare/v0.0.6-beta...v0.0.9-beta) (2023-03-27)


### Bug Fixes

* fix ip address bug ([33885c5](https://github.com/thomas-void0/fool-deploy/commit/33885c5f9588eb000e051cf40aca775416d25eef))
* fix packagecommand get error ([b9a2ee6](https://github.com/thomas-void0/fool-deploy/commit/b9a2ee6cae3740808b109a270ecabab218a6d1f6))


### Features

* add port tip ([bbf73dc](https://github.com/thomas-void0/fool-deploy/commit/bbf73dc13db6797dbea459a0fd001a96c899322f))
* add probe port ([cfc7080](https://github.com/thomas-void0/fool-deploy/commit/cfc7080682e663a00f046718aafeac87eae9fbaf))
* add visit address console ([c2d92cf](https://github.com/thomas-void0/fool-deploy/commit/c2d92cf4d35bb53558d1a3d2c606e4edabbaa13c))



## [0.0.6-beta](https://github.com/thomas-void0/fool-deploy/compare/v0.0.5-beta...v0.0.6-beta) (2023-03-26)


### Bug Fixes

* fix output generate bug ([c699cb8](https://github.com/thomas-void0/fool-deploy/commit/c699cb89d252269aabe7246fbb059d9f6d70a273))



## [0.0.5-beta](https://github.com/thomas-void0/fool-deploy/compare/v0.0.4-beta...v0.0.5-beta) (2023-03-26)


### Features

* add output field ([721041e](https://github.com/thomas-void0/fool-deploy/commit/721041e88871661725736b511b79bf79345f9854))



## [0.0.4-beta](https://github.com/thomas-void0/fool-deploy/compare/v0.0.3-beta...v0.0.4-beta) (2023-03-26)


### Bug Fixes

* fix node run bug ([5bf859e](https://github.com/thomas-void0/fool-deploy/commit/5bf859e35cda53dc9a2a6a68b46aa64b858b0df7))



## [0.0.3-beta](https://github.com/thomas-void0/fool-deploy/compare/v0.0.2-beta...v0.0.3-beta) (2023-03-26)



## [0.0.2-beta](https://github.com/thomas-void0/fool-deploy/compare/v0.0.1-beta...v0.0.2-beta) (2023-03-26)



## [0.0.1-beta](https://github.com/thomas-void0/fool-deploy/compare/2d3a1b29adc2604a768255f3b3313deef0bc1572...v0.0.1-beta) (2023-03-26)


### Bug Fixes

* fix dir delete bug ([71ae165](https://github.com/thomas-void0/fool-deploy/commit/71ae165074f1bc1b6ffa25ff956342889ae51634))
* fix jest error ([ad204d8](https://github.com/thomas-void0/fool-deploy/commit/ad204d87f62b551b19664a3294fc9176a6f7145b))
* fix npm build error ([7a334ed](https://github.com/thomas-void0/fool-deploy/commit/7a334ed2b3f5409d3dffab43d14cf0ef48975b19))
* fix pnpm error for copy .npmrc file ([87ec3cb](https://github.com/thomas-void0/fool-deploy/commit/87ec3cb2ddc07a9ef2301e709b9ab4845342b271))
* fix zh.md link error ([70ba08a](https://github.com/thomas-void0/fool-deploy/commit/70ba08aaf7bcc79b34896ac8bd7c96bd123060fd))


### Features

* add create function ([70f8cb6](https://github.com/thomas-void0/fool-deploy/commit/70f8cb6ae544a4ea4a4bb59f4faed79a63054906))
* add crypto ([2d3a1b2](https://github.com/thomas-void0/fool-deploy/commit/2d3a1b29adc2604a768255f3b3313deef0bc1572))
* add readme ([1aa6f60](https://github.com/thomas-void0/fool-deploy/commit/1aa6f605b0b10a8fde98ed4989048cd13a497393))
* add shelljs ([a67abb9](https://github.com/thomas-void0/fool-deploy/commit/a67abb9f0004f5d337c876afa2d2cf4bd5775999))
* add success log ([79fcd2e](https://github.com/thomas-void0/fool-deploy/commit/79fcd2e3b20c32c59067586b5e897daf699ec485))
* add test ([255929e](https://github.com/thomas-void0/fool-deploy/commit/255929ef57706a21b628e6bea63f112ec0e958d4))
* generateNginxContent ([4355f78](https://github.com/thomas-void0/fool-deploy/commit/4355f78ed8fb641acae5aafdc86a7d6dd43297f4))





src/lib/generateOptions.ts
import fs from 'fs';
import { Options } from '../typings';
import getPackageCommand from './getPackageCommand';
import getValidPort from './getValidPort';

let options: Options = {
  cache: true,
  port: 2333,
  nodeVersion: '18.14-alpine',
  nginxVersion: '1.22.1',
  imageName: 'fool-deploy:prod',
  output: 'dist',
};

async function generateOptions() {
  const cwdPath = process.cwd();
  // read package.json for get name
  const pkgPath = `${cwdPath}/package.json`;
  if (fs.existsSync(pkgPath)) {
    const str = fs.readFileSync(pkgPath).toString('utf-8');
    const { name = 'fool-deploy' } = JSON.parse(str);
    options.imageName = `${name}:prod`;
  } else {
    throw new Error(
      `Error: ${cwdPath} not found package.json，shell must running in projcet root dir.`
    );
  }

  // read foolDeploy.json config
  const jsonPath = `${cwdPath}/.foolrc`;
  if (fs.existsSync(jsonPath)) {
    const str = fs.readFileSync(jsonPath).toString('utf-8');
    const externalOptions = JSON.parse(str);
    options = { ...options, ...externalOptions };
  }

  const { imageName, packageCommand, buildCommand } = options;

  // insure imageName is valid
  if (imageName.split(':').length < 2) {
    options.imageName = `${imageName}:prod`;
  }

  // insure packageName is valid
  if (!packageCommand) {
    options.packageCommand = getPackageCommand();
  }

  // insure buildCommand
  if (!buildCommand) {
    const { packageCommand } = options;
    options.buildCommand =
      packageCommand === 'npm'
        ? `${packageCommand} run build`
        : `${packageCommand} build`;
  }

  // insure port is valid
  options.port = await getValidPort(options.port);

  return options;
}

export default generateOptions;
